File: e.VOC.us/Scripts/TypeScript/Keyboard.ts
///<reference path='GameObject.ts' />

module eVOCus {
    export class Keyboard {
        keysDown: number[] = [];
        keysPressed: number[] = [];
        constructor() {
            document.addEventListener('keydown', this.keyDown.bind(this), false);
            document.addEventListener('keyup', this.keyUp.bind(this), false);
        }

        update() {
            this.keysPressed = [];
        }

        private keyDown(event: KeyboardEvent) {
            if (this.keysDown.indexOf(event.keyCode) == -1) {
                this.keysDown.push(event.keyCode);
                Game.game.hub.server.keyboardInput(event.keyCode, "down");
            }
        }

        private keyUp(event: KeyboardEvent) {
            Game.game.hub.server.keyboardInput(event.keyCode, "up");
            this.keysPressed.push(event.keyCode);
            var index = this.keysDown.indexOf(event.keyCode, 0);
            if (index != undefined) {
                this.keysDown.splice(index, 1);
            }
        }

        isKeyDown(key: number): boolean {
            return this.keysDown.indexOf(key) != -1;
        }

        isKeyPressed(key: number): boolean {
            return this.keysPressed.indexOf(key) != -1;
        }
    }
}

File: e.VOC.us/Game/Keyboard.cs
using System.Collections.Generic;
using Newtonsoft.Json;

namespace e.VOC.us.Game
{
    public class Keyboard
    {
        public List<int> KeysDown = new List<int>();
        public  List<int> KeysPressed = new List<int>();

        public void Update()
        {
            KeysPressed.Clear();
        }

        public bool IsKeyDown(int key)
        {
            return KeysDown.Contains(key);
        }

        public bool IsKeyPressed(int key)
        {
            return KeysPressed.Contains(key);
        }
    }
}